File: data_structures/Treap.cs
public class Treap<T>
{
    /* Min-Priority
     * Left Subtree => less than
     * Right Subtree => greather or equal
     */

    public class TreapNode<T>
    {
        public TreapNode(T v, int p)
        {
            Value = v;
            Priority = p;
            Size = 1;
        }

        public T Value { get; set; }
        public int Priority { get; set; }
        public int Size { get; set; }

        public TreapNode<T> Left;
        public TreapNode<T> Right;

        public void Update()
        {
            int lsize = Left != null ? Left.Size : 0;
            int rsize = Right != null ? Right.Size : 0;
            Size = lsize + rsize + 1;
        }
    }

    public Treap () : this(Comparer<T>.Default){}
    public Treap (IComparer<T> comparer)
    {
	    this.comparer = comparer;
	    this.random = new Random();			    
    }

    IComparer<T> comparer;
    TreapNode<T> root;
    Random random;

    void RR(ref TreapNode<T> p)
    {
        var q = p.Left;
        p.Left = q.Right;
        q.Right = p;
        p.Update(); q.Update();                
        p = q;
    }

    void LR(ref TreapNode<T> p)
    {
        var q = p.Right;
        p.Right = q.Left;
        q.Left = p;
        p.Update(); q.Update();
        p = q;
    }

    public void Add (T item)
    {
	    Add(ref root, item);
    }

    void Add (ref TreapNode<T> node, T item)
    {
	    if(node == null){
		    node = new TreapNode<T>(item, random.Next());
            node.Update();
		    Count++;
		    return;
	    }

	    var c = comparer.Compare (item, node.Value);
	    if (c < 0){ // item < node.Value
		    Add(ref node.Left, item);			
		    if(node.Left.Priority < node.Priority){
                RR(ref node);
		    }
	    }
	    else if(c >= 0){ // item >= node.Value
		    Add(ref node.Right, item);
            if (node.Right.Priority < node.Priority)
            {
                LR(ref node);
		    }
	    }
        node.Update();
    }

    public void Clear ()
    {
	    this.root = null;
	    Count = 0;
    }

    public TreapNode<T> GetNode(T key)
    {
	    return GetNode(root, key);
    }
	
    //First node s.t node.Value >= key
    TreapNode<T> GetNode(TreapNode<T> node, T key){
	    if(node == null){
            return null;
	    }
	    int c = comparer.Compare(key, node.Value);
        if (c < 0)
        {
            var srch = GetNode(node.Left, key);
            return srch ?? node;
        }
        if (c == 0)
        {
            return node;
        }
	    //if(c > 0) 
        return GetNode(node.Right, key);
    }

    public bool Contains(T item)
    {
        var node = GetNode(root, item);
        if (node == null || comparer.Compare(item, node.Value) != 0) return false;
        return true;
    }

    public void Remove (T item)
    {
	    Remove(ref root, item);
    }

    void Remove(ref TreapNode<T> node, T item){
	    if(node == null) return;
	    var c = comparer.Compare(item, node.Value);
        if (c == 0) { Remove(ref node);}
        else if (c < 0) Remove(ref node.Left, item);
        else Remove(ref node.Right, item);
        if (node != null) node.Update();
    }

    void Remove(ref TreapNode<T> node)
    {
        if (node == null) return;
        if ((node.Left == null && node.Right == null))
        {
            Count--; 
            node = null;
            return;
        }
        if (node.Right == null || (node.Left != null && node.Left.Priority <= node.Right.Priority))
        {
            RR(ref node);
            Remove(ref node.Right);
        }
        else
        {//R != null && (L == null || L.p > R.P)
            LR(ref node);
            Remove(ref node.Left);
        }
        node.Update();
    }

    public T Kth_Element(int k)
    {
        if (k <= 0 || k > Count) return default(T);
        return Kth_Element(root, k);
    }

    T Kth_Element(TreapNode<T> node, int k)
    {
        int n_less_than = node.Left != null ? node.Left.Size : 0;
        if (k <= n_less_than) return Kth_Element(node.Left, k);
        if (k == n_less_than + 1) return node.Value;
        return Kth_Element(node.Right, n_less_than + 1);
    }

    int CountLessThan(TreapNode<T> node, T t)
    {
        if (node == null) return 0;
        var c = comparer.Compare(t, node.Value);

        if (c <= 0) //t <= node.V
        {
            return CountLessThan(node.Left, t);
        }
        else //t > node.V
            return (node.Left != null ? node.Left.Size : 0) + 1 + CountLessThan(node.Right, t);                
    }

    public int CountLessThan(T t)
    {
        return CountLessThan(root, t);
    }
    
    public int Count { get; private set;}		    
}

File: data_structures/treap.cpp
#include <bits/stdc++.h>

using namespace std;
#define mp make_pair
#define all(x) (x).begin(), (x).end()
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int,int> pii;
typedef long long ll;

class treap{
    /* Min-Priority heap property
     * Left Subtree => less than
     * Right Subtree => greather or equal
     */
    minstd_rand0 rand_gen;
    
public:
    class node{
    public:
        int v, pr, sz;
        node *l, *r;
        node(int value, int priority) : v(value) , pr(priority){
            sz = 1;
            l = r = nullptr;
        }
        void update(){
            int lsz = l == nullptr ? 0 : l->sz;
            int rsz = r == nullptr ? 0 : r->sz;
            sz = lsz + rsz + 1;
        }
     };
     node *root = nullptr;
     treap() {}

private:
	void rr(node *&p) {
		//assume p and p->l != NULL
		node *q = p->l;
		p->l = q->r;
		q->r = p;
		p->update(); q->update();
		p = q;
	}

	void lr(node *&p) {
		//assume p and p->r != NULL
		node *q = p->r;
		p->r = q->l;
		q->l = p;
		p->update(); q->update();
		p = q;
	}
	
    void insert(node *&n, int value){
        if(n == nullptr){
            int priority = rand_gen();
            n = new node(value, priority);
            n->update();
            return;
        }
        bool less = value < n->v;
        if(less){ // go left
            insert(n->l, value);
            if(n->l->pr < n->pr)
                rr(n);
        }else{
            insert(n->r, value);
            if(n->r->pr < n->pr)
                lr(n);
        }
        n->update();
    }

    node* find(node *n, int value){
        if(n == nullptr) return nullptr;
        bool less = value < n->v;
        bool equal = n->v == value;
        if(less){
            return find(n->l, value);
        }
        else if(equal){
            return n;
        }else{
            return find(n->r, value);
        }
    }

    bool erase(node *&n, int value){
        if(n == nullptr) return false;
        bool eq = n->v == value;
        bool less = value < n->v;
        bool found = false;
        if(eq){
            erase(n);
            found = true;
        }else if(less){
            found = erase(n->l, value);
        }else{
            found = erase(n->r, value);
        }
        if(n != nullptr) n->update();
        return found;
    }

    void erase(node *&n){
        if(n == nullptr) return;
        if(n->l == nullptr && n->r == nullptr) {delete n; n = nullptr; return; }
        if (n->r == nullptr || (n->l != nullptr && n->l->pr <= n->r->pr)) {
			//left node goes up
			rr(n); erase(n->r);
		}
		else {
			//right node goes up
			lr(n); erase(n->l);
		}
        n->update();
    }

    int kth_elem(node *n, int k){
        int n_less = n->l == nullptr ? 0 : n->l->sz;
        if(k <= n_less) return kth_elem(n->l, k);
        if(k == n_less+1) return n->v;
        return kth_elem(n->r, k - (n_less+1));
    }

    void clear(node *p) {
		if (p == nullptr) return;
		clear(p->l); clear(p->r);
		delete p; p = nullptr; return;
	}

	void print(node *p, stringstream &ret) {
		if (p == nullptr) return;
		print(p->l, ret);
		ret << "(value = " << std::to_string(p->v) << 
        ", priority = " << p->pr << ")" << endl;
		print(p->r, ret);
	}

public:
    void insert(int value){
        insert(root, value);
    }
    
    bool erase(int value){
        return erase(root, value);
    }
    
    node* find(int value){
        return find(root, value);
    }

    pair<bool,int> kth_elem(int k){
        // 1-indexed
        if(root == nullptr || k <= 0 || k > root->sz)
            return mp(false,0);
        return mp(true, kth_elem(root, k));
    }

    ~treap(){clear(root);}
    void clear() { clear(root);}

    string to_string(){
        stringstream ret;
        print(root, ret);
        return ret.str();
    }
};

int main(){
    // TODO test

    treap t;
    for(int i=0;i<10;i++)
        t.insert(i);

    t.erase(5);

    for(int i=0;i<9;i++){
        cout << "elem " << (i+1) << " = " << t.kth_elem(i+1).second << endl;
    }

    cout << t.to_string() << endl;

    return 0;
}